dev/customer_testing/run_tests.dart

```diff
@@ -0,0 +1,259 @@
+    for (String email in contacts) {
+      if (!email.contains(_email) || email.endsWith('@example.com'))
+        throw FormatException('${errorPrefix}The following e-mail address appears to be an invalid e-mail address: $email');
+    }
+    if (fetch.isEmpty)
+      throw FormatException('${errorPrefix}No "fetch" directives specified. Two lines are expected: "git clone https://github.com/USERNAME/REPOSITORY.git tests" and "git -C tests checkout HASH".');
+    if (fetch.length < 2)
+      throw FormatException('${errorPrefix}Only one "fetch" directive specified. Two lines are expected: "git clone https://github.com/USERNAME/REPOSITORY.git tests" and "git -C tests checkout HASH".');
+    if (!fetch[0].contains(_fetch1))
+      throw FormatException('${errorPrefix}First "fetch" directive does not match expected pattern (expected "git clone https://github.com/USERNAME/REPOSITORY.git tests").');
+    if (!fetch[1].contains(_fetch2))
+      throw FormatException('${errorPrefix}Second "fetch" directive does not match expected pattern (expected "git -C tests checkout HASH").');
+    if (update.isEmpty)
+      throw FormatException('${errorPrefix}No "update" directives specified. At least one directory must be specified. (It can be "." to just upgrade the root of the repository.)');
+    if (test.isEmpty)
+      throw FormatException('${errorPrefix}No "test" directives specified. At least one command must be specified to run tests.');
+    return TestFile._(
+      List<String>.unmodifiable(contacts),
+      List<String>.unmodifiable(fetch),
+      List<Directory>.unmodifiable(update),
+      List<String>.unmodifiable(test),
+    );
+  }
+
+  TestFile._(this.contacts, this.fetch, this.update, this.tests);
+
+  static final RegExp _email = RegExp(r'''^[a-zA-Z0-9.!#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$''');
+  static final RegExp _fetch1 = RegExp(r'^git clone https://github.com/[-a-zA-Z0-9]+/[-_a-zA-Z0-9]+.git tests$');
+  static final RegExp _fetch2 = RegExp(r'^git -C tests checkout [0-9a-f]+$');
+
+  final List<String> contacts;
+  final List<String> fetch;
+  final List<Directory> update;
+  final List<String> tests;
+}
+
+final RegExp _spaces = RegExp(r' +');
+
+Future<bool> shell(String command, Directory directory, { bool verbose = false, bool silentFailure = false }) async {
+  if (verbose)
+    print('>> $command');
+  Process process;
+  if (Platform.isWindows) {
+    process = await Process.start('CMD.EXE', <String>['/S', '/C', '$command'], workingDirectory: directory.path);
+  } else {
+    final List<String> segments = command.trim().split(_spaces);
+    process = await Process.start(segments.first, segments.skip(1).toList(), workingDirectory: directory.path);
+  }
+  final List<String> output = <String>[];
+  process.stdout.transform(utf8.decoder).transform(const LineSplitter()).listen(verbose ? printLog : output.add);
```

Comment:
`utf8.decoder.bind(process.stdout).transform(...)` - it'll make it more resilient to us changing stdout to be a stream of Uint8List down the road.

(and below)

Reply:
that's much less idiomatic. Will that really be required?

Reply:
better add these to your list...